### ISSUE TITLE
Feature: Ability to clear log history

### ISSUE THREAD

**Issue by djessup** [docker/compose#1083]
A feature I've thought would be useful since originally using Fig, and now Compose would be the ability to clear the log history for Composed-managed containers. Long-running or "chatty" containers can end up with a lot of log noise that may not be wanted.

I'd expect a command like the following would solve the problem:
```$ docker-compose logs --clear [service]```

**Comment by dnephin** [comment]
I don't think that is a feature supported by the docker daemon.  Looking at the api docs, the only option would be to truncate the returned logs to a limited number of lines per container:  

https://docs.docker.com/reference/api/docker_remote_api_v1.17/#get-container-logs

**Comment by metal3d** [comment]
:+1: for this issue. I actually use compose to develop a website in Golang, Mongodb and nginx... 5 days after begining I have a long logs that becomes to be worrying. Each time I restart containers append many lines to the log.
@dnephin I don't understand if you give a solution (that I don't understand :)) or if you propose to check if it's possible with api. Sorry for my bad english.

**Comment by thaJeztah** [comment]
Docker 1.6 will add support for logging drivers, see https://github.com/docker/docker/pull/10568 (currently; JSON, syslog and "none") work is in progress for basic log-rotation; https://github.com/docker/docker/pull/11485

**Comment by metal3d** [comment]
Good to hear that, thanks a lot :)

**Comment by djessup** [comment]
Thanks for the background and update on Docker 1.6 - looking forward to it!

**Comment by sgarbesi** [comment]
`docker logs -c (clear) <container>` would be great.

+1

**Comment by baptistedonaux** [comment]
+1 Really important

**Comment by egyptianbman** [comment]
Holy cow, I just sat through several minutes of logs to get to the end.  I'd really appreciate this as well for the sake of not having to rebuild containers constantly.

**Comment by jghibiki** [comment]
+1

**Comment by vronin** [comment]
+1

**Comment by junajan** [comment]
+1

**Comment by knutole** [comment]
:+1:

**Comment by Rodeoclash** [comment]
+1

**Comment by nafg** [comment]
+2

**Comment by FnTm** [comment]
+1

**Comment by dmage** [comment]
Just for the record, with Docker 1.8 and docker-compose 1.4 there is already exists a method to limit log size using https://docs.docker.com/compose/yml/#log-driver and log-opt max-size:
```
  log_driver: "json-file"
  log_opt:
    max-size: "100k"
    max-file: "20"
```

**Comment by knutole** [comment]
@dmage Thanks, exactly what I need.

**Comment by OlivierCuyp** [comment]
+1000

**Comment by sheyman** [comment]
+1 would LOVE this

**Comment by joegoggins** [comment]
+1

**Comment by Rodeoclash** [comment]
@dmage solution works great for us. Those others who are +1ing this, what is wrong with the solution that he's given?

**Comment by rosskevin** [comment]
@Rodeoclash - I think one use case for this is reuse of a given set of containers for a new run i.e. CI test run. The old logs are irrelevant to the new run, so a `clear` prior to the next command would clear up confusion.

**Comment by SpeCT** [comment]
+1

**Comment by ballinette** [comment]
+1, same need as @rosskevin

**Comment by dnephin** [comment]
Ok, to recap:

* you can do this from a compose file (see https://github.com/docker/compose/issues/1083#issuecomment-141936600)
* #265 covers being able to limit the output from the `logs` command

Going to close this since it's already support or tracked in other issues.

**Comment by chris-martin** [comment]
I don't quite understand why this was closed. How do you clear the log history?

**Comment by egyptianbman** [comment]
I add:
```
  log_opt:
    max-size: 50k
```
To limit the length of the logs.

**Comment by sgarbesi** [comment]
They really need to add this, it's essential. Limiting logs is good and all but there should be a simple command to clear logs.

```bash
docker logs -c <container>
```

What gives?

**Comment by chris-martin** [comment]
I understand how to limit the log size, but how do you *clear* the log?

**Comment by dnephin** [comment]
I don't think clearing a log is supported by the docker engine, which is what managed the logs.

**Comment by geekscrapy** [comment]
+1 for being able to flush the logs...

**Comment by kongchen** [comment]
+1

**Comment by ObjReponse** [comment]
+1

**Comment by kierr** [comment]
+1

**Comment by mjaverto** [comment]
+1

**Comment by helinwang** [comment]
+1 for clear log command.

**Comment by dgommeren** [comment]
+1 for clear log command

**Comment by max3-05** [comment]
+1

**Comment by Jarema** [comment]
+1

**Comment by mskutin** [comment]
+1

**Comment by kytwb** [comment]
:+1:

**Comment by emanuil** [comment]
+1

**Comment by mcuadros** [comment]
+1

**Comment by hosszukalman** [comment]
I usually clear the logs with `docker-compose up --force-recreate`.

**Comment by wavenator** [comment]
Adding a flag to the docker command line would be great too.

`docker --max-log-size 50k`

**Comment by sgarbesi** [comment]
**docker-logs-clean.sh**

```bash
#!/bin/bash

rm $(docker inspect $1 | grep -G '"LogPath": "*"' | sed -e 's/.*"LogPath": "//g' | sed -e 's/",//g');
```

---

**Invocation:**

```bash
sudo ./docker-logs-clean.sh <container-name>;
```

**Comment by CowPanda** [comment]
@sgarbesi thanks!
+1

**Comment by ariden** [comment]
+1 for clear log command

**Comment by paolomainardi** [comment]
Closed in favor of ?

**Comment by dnephin** [comment]
To reiterate my comment from https://github.com/docker/compose/issues/1083#issuecomment-149357280:

Ok, to recap:

* you can do this from a compose file (see https://github.com/docker/compose/issues/1083#issuecomment-141936600)
* #265 covers being able to limit the output from the logs command
* #1756 covers the re-used container case

Going to close this since it's already supported or tracked in other issues.

If what you're looking for is some command like`docker logs --clear`, that is not supported by docker engine, so you would need to request that on docker/docker. However, I think the above options should already be sufficient for most cases. What most people want is to just show a subset of the logs, not actually remove them.

**Comment by djessup** [comment]
Thanks @dnephin, and everyone else for their contributions. I raised this issue almost a year ago now, and judging from the comments since then it seems pretty clear to me that log management is a pain point for a lot of compose users.

There have been a few workarounds mentioned, myself I mostly use `max-size` to keep the logs to a reasonable length. This helps a lot, and I'm grateful for these workarounds, but it's important to keep in mind this is a _workaround_, not a solution.

It's also clear to me now that part of the responsibility for fixing this lies within the Docker logging system and that it would need to provider a `clear` command for Compose to utilise - fair enough.

All that said however, there are a few features that have made it into Docker releases since this ticket was created - namely `--since=<timestamp>` and `--tail=<num-lines>` that could be supported by Compose to come closer to giving a real solution to this issue.

For example, supporting `--since` could make something like this possible:
```
$ docker-compose logs --since=now my_container
```
or
```
$ docker-compose logs --since=5m my_container
```
Support for `--tail` would also be useful, e.g.
```
$ docker-compose logs --tail=100 my_container
```
and of course, combinations thereof. It may also make sense to support these in `docker-compose.yml` as part of the `logging` block, but even without that just by supporting these two options I suspect you could satisfy the majority of people who've +1'd this ticket.

In summary, thanks again for all the feedback and workarounds, and for Docker and Compose themselves - they're both great products - I hope you'll consider the ideas put forward in this thread and continue to make these products even greater.

**Comment by dnephin** [comment]
See #2227 for a summary

**Comment by djessup** [comment]
:+1: thanks @dnephin, looking forward to it!

**Comment by samarti** [comment]
What I did was to manually delete the ``` <container-id>-json.log``` files placed inside ```/var/lib/docker/containers/<container-id>/``` (use ```sudo```). Once I run ```docker-compose logs```, the log was empty. It's not a solution but with a proper ```.bash``` file you can automatize the cleaning before each build.

**Comment by douineauromain** [comment]
+1

**Comment by godely** [comment]
+1

**Comment by ZakariaBoualaid** [comment]
+1

**Comment by cmtec** [comment]
+1

**Comment by vhosakot** [comment]
+2

**Comment by joshbedo** [comment]
+2

**Comment by eexit** [comment]
+1

**Comment by vivex** [comment]
+20

**Comment by iagopiimenta** [comment]
+1

**Comment by JohnWelle** [comment]
+1

**Comment by uidzero** [comment]
+1

**Comment by pbn04001** [comment]
+1

**Comment by kassanmoor** [comment]
@sgarbesi After run that clean command, the log feature will work normally?

**Comment by qrpike** [comment]
+1

**Comment by lvitals** [comment]
docker-logs-clean.sh

```bash
#!/bin/bash

for container_id in $(docker ps -a --filter="name=$name" -q);

    do file=$(docker inspect $container_id | grep -G '"LogPath": "*"' | sed -e 's/.*"LogPath": "//g' | sed -e 's/",//g');

    if [ -f $file ]
      then
          rm $file;
    fi

done
```

**Comment by vinogradov** [comment]
+1 for command line option for clearing logs manually

**Comment by sgarbesi** [comment]
@kassanmoor Does for me.

**Comment by Nicolas-Leucci** [comment]
+1

**Comment by mirkods** [comment]
👍

**Comment by HerbertChu** [comment]
+1 👍

**Comment by krzysztof-kabala** [comment]
+1 !

**Comment by dottgonzo** [comment]
+1

**Comment by paustin01** [comment]
+1

**Comment by zfrenchee** [comment]
+1

**Comment by wazoo** [comment]
This would be a great feature to have, I have a few containers that are java based and after even a day or two or when troubleshooting it can take 30 seconds or more to do a `docker logs -f <container>` because Java loves those multi-line logs.  

Another thing that might be easier to implement is another flag on logs to start a tail without echoing all of the existing logs `docker logs -f -n <container>` for example (meaning follow only new logs).  This would only echo log messages received after running the command.  

The opposite would also work (and would be closer to the way that gnu tail works) is doing a -f echos the last 5-10 lines and then new lines by default and adding a flag to echo everything as it does now (maybe `docker logs -f -a <container>` or something).

On that note I would love to have the feature from gnu tail that allows specification of how many lines to tail like `docker logs -100 <container>` gives me the last100 lines.

**Comment by kopax** [comment]
+1

**Comment by supremekai** [comment]
+1

**Comment by michaelfreund** [comment]
+1

**Comment by ismay** [comment]
+1

**Comment by bjspatel** [comment]
+1

**Comment by fatcloud** [comment]
+1

**Comment by pvoisin** [comment]
+1

**Comment by Valten1992** [comment]
+1

**Comment by twooster** [comment]
+1

**Comment by shuklara** [comment]
+1

**Comment by ahfeel** [comment]
Is there a trick for Docker for mac ? Thanks and +1 for a command line option for this :)

**Comment by alirezaDavid** [comment]
+1

**Comment by 8026787** [comment]
+1

**Comment by Manoj-nathwani** [comment]
+1

**Comment by DavidPesticcio** [comment]
Hi all,

Because deleting files that could be open is generally a BAD idea - especially if you are trying to save diskspace! - I have expanded upon the initial efforts of @sgarbesi & @lvitals (thanks for the ideas guys) to produce the slightly more functional script below.

Copy the code below into a file
`vi ./docker-log-clear.sh`
Make the file executable
`chmod +x ./docker-log-clear.sh`
Then execute with
`sudo ./docker-log-clear.sh`

Uncomment the line with the `>` when you trust the script... that's where the magic happens. :-)


Feedback welcome.
Thanks.

PS. Tested, but not battle-hardened. Use at own risk.

```
#!/bin/bash

# NOTES:
#  Requires jq (https://github.com/stedolan/jq)
#  Does NOT delete logfile (BAD IDEA) - simply zeros file with redirect.
#  Handles single/all-running/all-existing containers - see end of script for usage.
#  Enjoy :-)

_get_container_logfile() {

  case $1 in

    running) _clear_container_logfile "$(docker ps -q)"
             ;;

        all) _clear_container_logfile "$(docker ps -aq)"
             ;;

          *) _clear_container_logfile "$(docker ps -a | awk -v ID=$1 '$1 ~ ID || $NF ~ ID {print $1}')"
             ;;

  esac

}


_clear_container_logfile() {

  if [ -z $1 ]
  then
    echo "Container name/id missing!"
    exit -1
  else
    for container in $1
    do
      logfile="$logfile $(docker inspect $container | jq -r '.[].LogPath')"
      echo "Clearing: $logfile"

      # Uncomment the next line when you trust the script!
      # > $logfile
    done
  fi

}


if [ -a "$(which jq)" ]
then
  case $1 in
    --clear) if [ -z $2 ]
             then
               echo "Container name/id missing!"
               exit -1
             else
               _get_container_logfile $2
             fi
             ;;

    --clear-running) _get_container_logfile running
                     ;;
 
    --clear-all) _get_container_logfile all
                 ;;

     *) echo "Usage:"
        echo "  --clear {container}    Clear logfile for a single container"
        echo "  --clear-running        Clear logfile for all running containers"
        echo "  --clear-all            Clear logfile for all containers"
        exit -1
        ;;
  esac
else
  echo "Requires \"jq\""
  exit -1
fi
```

**Comment by babasaikiran** [comment]
+1

**Comment by xiemeilong** [comment]
+1

**Comment by JonasT** [comment]
So since it was expressed multiple times that there is a wish for an explicit clear command and since there was never one added, any chance this will be reopened?

It depends on a docker engine change of course, but it's still something that will eventually need to be addressed in docker-compose as well - and it's definitely not fixed as far as an explicit command is concerned.

**Comment by gingerlime** [comment]
@DavidPesticcio I'm getting this error when running your script: `line 53: $logfile: ambiguous redirect` (after removing the comment)

**Comment by DavidPesticcio** [comment]
Hi @gingerlime, Seems like $TEMP is not being populated... maybe you don't have "mktemp" installed, or it's not in your path perhaps? :-/

"Works fine for me" - yeah I know, that's not much help to you, but it's true... :-)

I have updated the script, so you don't need "jq" any more - perhaps I should add a bail-out if mktemp is missing too... I thought it was a standard tool - but maybe you're not running the script within a "standard" install - like from within a minimal container maybe? :-)

Hope that helps!

**Comment by gingerlime** [comment]
I do have `mktemp`, and use it frequently...Didn't really spend much time debugging it. Ended up with the simpler script above that just nukes those logs. In our dev environment they are not important.

**Comment by fruch** [comment]
on my setup (i.e. running as no root user) this bash script isn't helping casue i'm getting premission denied trying to open the log file.

it's a bit weird, that as a user i can start docker, but can tocuh it's log files...
more of a reason to have something expend to the docker/docker-compose commands

**Comment by gamma** [comment]
I had to modify the script of @DavidPesticcio because of the errors mentioned before ... here it goes:

```
#!/bin/bash

# NOTES:
#  Does NOT delete logfile (BAD IDEA) - simply trims file with redirect.
#  Handles single/all-running/all-existing containers - see end of script for usage.
#  Enjoy :-)


_get_container_logfile() {

  case $1 in

    running) _trim_container_logfile "$(docker ps -q)" $2
             ;;

        all) _trim_container_logfile "$(docker ps -aq)" $2
             ;;

          *) _trim_container_logfile "$(docker ps -a | awk -v ID=$1 '$1 ~ ID || $NF ~ ID {print $1}')" $2
             ;;

  esac

}


_trim_container_logfile() {

  TEMP=$(mktemp)

  case $2 in
    *[!0-9]*) echo "[lines] must be a number - \"$2\" is not a number."
              exit 1
              ;;
        ''|*) MAX=${2:-1000}
              ;;
  esac


  if [ -z "$1" ]
  then
    echo "Container name/id unknown!"
    exit 1
  else
    for container in $1
    do
      logfile="$(docker inspect --format '{{.LogPath}}' $container)"
      if [ ! -f "$logfile" ]; then continue; fi
      echo "Keeping $MAX lines: $logfile"

      tail -n ${MAX} "$logfile" > "$TEMP"
      # Uncomment the next line when you trust the script!
      # cat "$TEMP" > "$logfile"
    done
  fi

  rm "$TEMP"
}


if [ -a "$(which docker)" ]
then
  case $1 in
    --trim) if [ -z $2 ]
            then
              echo "Container name/id missing!"
              exit 1
            else
              _get_container_logfile $2 $3
            fi
            ;;

    --trim-running) _get_container_logfile running $2
                    ;;

    --trim-all) _get_container_logfile all $2
                ;;

    *) echo "Usage:"
       echo "  --trim {container} [lines]   Keep [lines] of logfile for a single container"
       echo "  --trim-running     [lines]   Keep [lines] of logfile for all running containers"
       echo "  --trim-all         [lines]   Keep [lines] of logfile for all containers"
       echo "Default: lines=1000"
       exit 1
       ;;
  esac
else
  echo "Requires \"docker\""
  exit 1
fi
```

**Comment by JonasT** [comment]
@dnephin any chance this could be reopened, since there is an obvious demand to have an explicit one-time clear command for logs?

**Comment by oogali** [comment]
Truncate the logs for a given container (must be root):

    cp /dev/null $(docker inspect -f '{{.LogPath}}' container_name)

You want truncation, not deletion. (Deleting the file referenced by an open file handle doesn't reclaim space until the process, in this case, the Docker daemon, completely exits)

**Comment by JonasT** [comment]
@oogali fair enough. Still would be nice to have a proper command for it.

**Comment by fx-dx** [comment]
+1

**Comment by lpoulter** [comment]
+1

**Comment by fx-dx** [comment]
+1

**Comment by fx-dx** [comment]
would be nice to have a proper command for it

**Comment by j2l** [comment]
+1
would be nice to have a proper command for it to do it on demand.

**Comment by nickwinn** [comment]
This is a no brainer, setting a start date as a filter is more work then just clearing it to look at current output. Please add this feature.

**Comment by erikverheij** [comment]
+1

**Comment by MichaelErmer** [comment]
+1

**Comment by heytrav** [comment]
+1

**Comment by piotr-s-brainhub** [comment]
+1

**Comment by eduardoleal** [comment]
+1

**Comment by Joshfindit** [comment]
Tangentially related:

When doing `docker-compose logs -f`, automatically default to `--tail=30` (or any other number that's reasonable)

**Comment by piotr-s-brainhub** [comment]
`docker-compose logs -f` is not enough because with large number of logs displaying it takes a very long time

**Comment by phlegx** [comment]
+1

**Comment by luchesar** [comment]
+1

**Comment by edwardmp** [comment]
+1

**Comment by rabahs** [comment]
+1

**Comment by htdttb** [comment]
+1

**Comment by atedja** [comment]
+1

`docker logs` is a trap now if you have containers with gajillion logs.

**Comment by crash21** [comment]
+1

**Comment by ahmetkucuk** [comment]
+1

**Comment by sychevsky** [comment]
+1

**Comment by OzzyCzech** [comment]
+1

**Comment by JeanLescut** [comment]
Or there is this workaround also 

`
logpath=`docker``` inspect --format='{{.LogPath}}' reveelium_metricsextraction_1` && mv $logpath $logpath".bckup"
`

**Comment by ReFruity** [comment]
Please, stop commenting +1. It makes this thread really hard to read and extract valuable information. There is a thumbs up button on OP post for this.

**Comment by avindra** [comment]
Nothing here works for me with Docker for mac. I've hacked together something that works based on reading this thread and the docker forums however.

The problem with D4M, is that on mac, you need to actually run the commands on the xhyve vm.  So here's what I came up with. Add [these two functions](https://github.com/avindra/dotfiles/blob/4efe5c37ca614fcf1cfc3faa0dd8b2913b5228f5/.bash_profile#L66-L80) to your `.bash_profile`.

Now, `docker-logs-clean` looks like this:

```bash
#!/bin/bash -e

if [[ -z $1 ]]; then
    echo "No container specified"
    exit 0
fi

logFile=$(docker inspect -f '{{.LogPath}}' $1 2> /dev/null)

echo -n "Cleaning ${logFile}... "
d4mexec << EOF
> $logFile
EOF
echo "done"
```

Note that I am not `rm`ing the log file, but just doing `>`, which will fully truncate the file.

**Comment by tgjamin** [comment]
FWIW in docker-compose version 2 the feature to limit log file sizes:

```
version: '2'
services:
  my-service:
    image: nginx:alpine
    restart: always
    logging:
      # limit logs retained on host to 25MB
      driver: "json-file"
      options:
        max-size: "500k"
        max-file: "50"
```
It is not documented on the docker website very well, and this may be useful to others.

**Comment by heetap** [comment]
+1

**Comment by sintetico82** [comment]
+1

**Comment by d13g0s0uz4** [comment]
+1

**Comment by MPJHorner** [comment]
+1

**Comment by gsugambit** [comment]
+1

**Comment by oliviertassinari** [comment]
+1

**Comment by educatedwarrior** [comment]
-1    This thread sucks.   How many times does a person have to ask for the feature to truncate logs and still get a response on how to limit logs.  NOT THE SAME THING

**Comment by educatedwarrior** [comment]
The original issue was never resolved and the issue was closed anyway.  Interesting.

**Comment by dhakehurst** [comment]
would still be nice to have a docker-compose logs --clean

**Comment by tobiasmuehl** [comment]
Found this on [StackOverflow](https://stackoverflow.com/a/43570083/2064544):

`truncate -s 0 /var/lib/docker/containers/*/*-json.log`

**Comment by MartinNiederl** [comment]
Any updates?

**Comment by FelixFurtmayr** [comment]
when there is a log, we should be able to clear it in a simple way.

would be really nice, if you could build the command into it. this seems not hard.
thousands of people would be very happy not having to scroll through everything.

**Comment by FelixFurtmayr** [comment]
If you use docker-compose, use **lazydocker** for viewing logs. Then there is then no need to delete logs.  You get live logs and start with the latest logs. This helps very much with debugging.

![grafik](https://user-images.githubusercontent.com/16761317/66265229-5fc0bc00-e813-11e9-9eb5-649d0b8f8797.png)

https://github.com/jesseduffield/lazydocker
You can install it with one line and it is a good monitoring tool. It is a pitty that people are not so reasonable to understand that a "delete log" function is a good thing.

But the workaround with lazy docker does it for me. Thank you jesseduffield for giving us the possibility to be lazy with your monitoring tool :-)
Maybe consider donating to lazydocker if this makes debugging/monitoring for you also easier.
And for you docker developers / maintainers: why is the interface from docker not like that?
Docker is pretty awesome, but have a look at lazydocker; there is room for improvement.

**Comment by GreatFireWall** [comment]
+2147483647

**Comment by kagansari** [comment]
+49324893

**Comment by bf** [comment]
This is such a central problem with UX of docker-compose, I wonder why it has not been resolved. Every time I run docker-compose up, some of the *new* logs are put on the top of *super old* logs from 5 starts ago. What is this?

**Comment by GF-Huang** [comment]
This issue should be re-open.

**Comment by ell1e** [comment]
I would love to use this as well. Why can't there be a simple `docker-compose logs --clean-logs <service>`?

**Comment by srinitude** [comment]
Echoing everyone here lol

**Comment by eon01** [comment]
A simple workaround for me is using `docker-compose up --force-recreate`

**Comment by srinitude** [comment]
@eon01 Awesome hack! Thanks for sharing

**Comment by GF-Huang** [comment]
7 years passed.

**Comment by reesericci** [comment]
Please add this!!!

**Comment by rentianxiang** [comment]
Please add this!!!

**Comment by julianoamg** [comment]
sudo sh -c "truncate -s 0 /var/lib/docker/containers/**/*-json.log"

**Comment by julianoamg** [comment]
You guys are all sucks, there is no such command for clear all log history from docker compose, but we had this on above, who does the magic for you.

**Create a simple Linux alias and voila.**

# cadl = clear all docker logs
`alias cadl 'sudo sh -c ""truncate -s 0 /var/lib/docker/containers/**/*-json.log'`

**Comment by PavelNiedoba** [comment]
It would make even more sense if you can
```
docker compose logs clear [all | since | keep lines xxx | you name it]
```
